slatedb/src/db.rs

```diff
@@ -212,69 +250,58 @@ impl DbInner {
-                    .sum::<usize>();
-                let imm_memtable_size = guard
-                    .state()
-                    .imm_memtable
-                    .iter()
-                    .map(|imm| {
-                        let metadata = imm.table().metadata();
-                        self.table_store.estimate_encoded_size(
-                            metadata.entry_num,
-                            metadata.entries_size_in_bytes,
-                        )
-                    })
-                    .sum::<usize>();
-                imm_wal_size + imm_memtable_size
-            };
-            if mem_size_bytes >= self.settings.max_unflushed_bytes {
-                let (wal_table, mem_table) = {
+                let wal_size = self.wal_buffer.estimated_bytes().await?;
+                let imm_memtable_size = {
                     let guard = self.state.read();
-                    (
-                        guard.state().imm_wal.back().map(|imm| imm.table().clone()),
-                        guard.state().imm_memtable.back().cloned(),
-                    )
+                    // Exclude active memtable to avoid a write lock.
+                    guard
+                        .state()
+                        .imm_memtable
+                        .iter()
+                        .map(|imm| {
+                            let metadata = imm.table().metadata();
+                            self.table_store.estimate_encoded_size(
+                                metadata.entry_num,
+                                metadata.entries_size_in_bytes,
+                            )
+                        })
+                        .sum::<usize>()
                 };
+                wal_size + imm_memtable_size
+            };
+
+            if mem_size_bytes >= self.settings.max_unflushed_bytes {
                 warn!(
                     "Unflushed memtable and WAL size {} >= max_unflushed_bytes {}. Applying backpressure.",
                     mem_size_bytes, self.settings.max_unflushed_bytes,
                 );
 
-                match (wal_table, mem_table) {
-                    (Some(wal_table), Some(mem_table)) => {
-                        tokio::select! {
-                            result = wal_table.await_durable() => {
-                                result?;
-                            }
-                            result = mem_table.await_flush_to_l0() => {
-                                result?;
-                            }
-                        }
+                self.flush_immutable_memtables().await?;
+
+                let await_flush_to_l0 = async {
+                    let imm = {
+                        let guard = self.state.read();
+                        guard.state().imm_memtable.back().cloned()
+                    };
+                    match imm {
+                        Some(imm) => imm.await_flush_to_l0().await,
+                        None => std::future::pending().await,
```

Comment:
I don't follow why this is here. Won't it leak memory since every time we apply back pressure, we'll create a new task that never ends?

Reply:
a good thing about rust's Future is that it's also managed in RAII, so i think it will not produce memory leak unless bind this task handle into something have a longer lifeime. (i believe it's actually a struct that manages a statemachine that does nothing 🤔, and it memory is also considered same as a stack allocated struct).